Microsoft SQL Server data source

marefr opened the pull request:
Continuing and finalizing work from PR #10093

* Merged the latest changes from master and fixed conflicts due to Grafana v5.0 and recent changes in postgres/mysql datasources.
* Added support for $__timeGroup macro function with fill support
* Added docker blocks for mssql and mssql_tests (grafana/fake-data-gen have been updated). 
* Added more integration tests
* Documentation will be handled in separate PR targeting docs 5.1 branch
* Note: No logo included right now. As soon as we have a valid logo we'll fix that
* Encrypt mssql connection password in database
* Removes dynamic construction of metric column and other columns. This seems like a niche feature which can be solved by defining multiple queries. In the future we'll probably want to add support for defining series name by alias field similar to how other datasources have solved that, e.g. prometheus.

Fake data:
* You'll find a dashboard.json in docker/blocks/mssql that you can import into Grafana which are using fake data from grafana/fake-data-gen when you run mssql docker-compose.

Integration tests:
* You'll find a dashboard.json in docker/blocks/mssql_tests/ that you can import into Grafana after you've created a mssql test datasource and run the integration tests.

Other:
* Fixed pre-commit so that gofmt only are run for files in pkg directory. Earlier it ran for vendor directory as well which could make you think that dep ensure did some strange things.

linuxchips committed:
add server only build target "build-srv"

linuxchips committed:
MSSQL Data Source

linuxchips committed:
add _tests for mssql data source

linuxchips committed:
updating the query editor's syntax highlighting mode to sqlserver

linuxchips committed:
adding mssql docs

linuxchips committed:
fixing a few fromattings

linuxchips committed:
added missing cases for DATETIME datatype

linuxchips committed:
adding support for sgl native time datatypes

linuxchips committed:
Merge branch 'master' into master

linuxchips committed:
Update built_in_plugins.ts

marefr committed:
Merge branch 'master' into mssql_datasource

marefr committed:
docker: add mssql block

marefr committed:
mssql datasource: additional data type tests

marefr committed:
mssql datasource: support for timeGroup macro function

marefr committed:
docker: add test dashboard for mssql for visualizing data generated by fake-data-gen

marefr committed:
docker: mssql and mssql tests blocks with common build context

marefr committed:
mssql datasource: wip

marefr committed:
Merge branch 'master' into mssql_datasource

xlson committed:
mssql: adds fill to timeGroup macro.

marefr committed:
mssql: add timeGroup integration test

xlson committed:
mssql: timeGroup fill support added.

marefr committed:
mssql: disable mssql integration tests per default

marefr committed:
docs: update mssql documentation

marefr committed:
mssql: update plugin.json, added description and name MSSQL -> Microsft SQL Server

marefr committed:
mssql: support money, smallmoney and decimal data types

marefr committed:
mssql: additional integration tests

Metric query of table having multiple value columns
Annotation query

marefr committed:
docker: add test dashboard for mssql tests based on test data generated by integration test

marefr committed:
mssql: skip rendering of mssql logos until we have a valid logo

marefr committed:
mssql: minor improvements of query editor help

marefr committed:
mssql: strip inkscape from logo

marefr committed:
mssql: add alternative logo without text

marefr committed:
Merge branch 'master' into mssql_datasource

marefr committed:
docs: mssql documentation will go into another branch targeting next minor version

marefr committed:
mssql: remove logos for now

marefr committed:
Merge branch 'master' into mssql_datasource

marefr committed:
mssql: update test dashboard

marefr committed:
mssql: update readme

marefr committed:
mssql: cleanup and minor changes

marefr committed:
docker: update test dashboard for mssql tests

marefr committed:
mssql: allow host without port and fallback to default port 1433

marefr committed:
mssql: cleanup and minor refactor

marefr committed:
test

marefr committed:
fix: only run gofmt on pkg directory omitting vendor directory

marefr committed:
fix: dep ensure. now without gofmt on ventor directory

xlson reviewed (approved):
Looks good.

xlson commented on docker/blocks/mssql/build/Dockerfile:
```diff
@@ -0,0 +1,5 @@
+FROM microsoft/mssql-server-linux:latest
```
2017-CU4 looks like the latest pinned version.

marefr replied:
Good catch :+1: Will fix

marefr commented on pkg/tsdb/mssql/mssql.go:
```diff
@@ -0,0 +1,326 @@
+			// construct the metric name
+			// if there is more than 3 columns (more than one value) and there is
+			// a metric column, join them to make the metric name
+			if metricIndex == -1 {
```
That's a good question. Maybe not since postgres doesn't have this functionality. Think we should try and solve this problem in UI in the future for all sql plugins so that you can use an alias field similar to how other data sources have solved this, i.e. influxdb/prometheus. I'll remove this functionality.

marefr committed:
docker: pin microsoft/mssql-server-linux to 2017-CU4 tag

marefr committed:
mssql: remove dynamic construction of metric column and other columns

This seems like a niche feature which can be solved by defining multiple queries. In the future
we'll probably add support for defining series name by alias field similar to how other datasources
have solved that, e.g. prometheus.

marefr committed:
Merge branch 'master' into mssql_datasource

daniellee commented on public/app/plugins/datasource/mssql/partials/config.html:
```diff
@@ -0,0 +1,40 @@
+		</div>
+		<div class="gf-form max-width-15">
+			<span class="gf-form-label width-7">Password</span>
+			<input type="password" class="gf-form-input" ng-model='ctrl.current.password' placeholder="password"></input>
```
Should use secureJsonData to encrypt the password like the Postgres data source does:

https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/postgres/partials/config.html#L24-L28

and decrypt it on the backend:

https://github.com/grafana/grafana/blob/master/pkg/tsdb/postgres/postgres.go#L49

marefr replied:
Done!

marefr committed:
mssql: encrypt password in database

marefr committed:
mssql: add integration test to verify stored procedure usage

svenklemm commented:
I think it would be nice if all sql datasources had somewhat consistent macros. Postgres doesnt require the ```AS time``` alias if you use ```$__timeGroup``` since the macro already generates it. I think it makes sense if SQL Server datasource would do the same. 
See https://github.com/grafana/grafana/issues/10074

marefr commented:
@svenklemm thanks for the comment. I totally agree, but we need to add some additional logic to render queries differently based on whether macro is used in select, group by or order by. Reason is that you cannot use `timestamp as time` in a group/order by statement in mssql. This however works in postgres. 

First step is to get mssql merged to master. Second, we want to refactor mysql a bit so that the implementation is similar to postgres and mssql. Third, we really want to try and refactor postgres, mysql and mssql since there are a lot of code duplication right now and they're all very similar.

svenklemm commented:
You cant do that in postgres either but the ```$__timeGroup``` macro is intended to be used as a column since the output is the timestamp value for grafana. In the group by you use backreferences to refer to the output of $__timeGroup like ```GROUP BY 1```. Changing it later would break queries people have written with the old version of the macro. Not sure if SQL Server supports backreferences though.

marefr commented:
Yes, see my updated comment. You cannot use backreference in mssql in group by expression which is a pity.

svenklemm commented:
Oh well that sucks. I guess not producing the alias is indeed the best way to do it for now until some other solution is found.

marefr commented:
Yes unfortunately. 

Maybe possible to do the changes later without breaking change if we can be backward compatible and handle the cases when there is a `AS column` after a macro function - shouldn't be that hard.

Thanks a lot for your review and wise thoughts @svenklemm

bergquist commented on package.json:
```diff
@@ -119,7 +119,7 @@
     ],
     "*.go": [
-      "gofmt -w -s",
+      "gofmt -w -s pkg",
```
I dont think this works as expected. husky adds the filepath to the end of this cmd. So this becomes 
`gofmt -w -s pkg /what/ever/files/you/changes.go`
I think a better fix would be to only check go files in `pkg/**.go`

marefr replied:
Good catch. Will fix in master straight away!

marefr replied:
Good catch. Will fix in master straight away!

bergquist replied:
No rush :)

daniellee commented:
Thanks @linuxchips for all your hard work - this is a fantastic feature for Grafana.
